File: src/main/scala/resonantinduction/mechanical/fluid/transport/TilePump.java
package resonantinduction.mechanical.fluid.transport;

import net.minecraft.block.material.Material;
import resonantinduction.core.prefab.node.NodePressure;
import resonantinduction.mechanical.energy.grid.TileMechanical;
import net.minecraft.tileentity.TileEntity;
import net.minecraftforge.common.util.ForgeDirection;
import net.minecraftforge.fluids.Fluid;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.fluids.FluidTankInfo;
import net.minecraftforge.fluids.IFluidHandler;
import resonant.api.IRotatable;
import universalelectricity.api.core.grid.INode;
import universalelectricity.core.transform.vector.Vector3;

public class TilePump extends TileMechanical implements IRotatable, IFluidHandler
{
	private final NodePressure pressureNode;

	public TilePump()
	{
		super(Material.iron);

        normalRender(false);
        isOpaqueCube(false);
        customItemRender(true);
        setTextureName("material_steel");
		pressureNode = new NodePressure(this)
		{
			@Override
			public int getPressure(ForgeDirection dir)
			{
				if (mechanicalNode.getPower() > 0)
				{
					if (dir == getDirection())
					{
						return (int) Math.max(Math.abs(mechanicalNode.getTorque() / 8000d), 2);
					}
					else if (dir == getDirection().getOpposite())
					{
						return (int) -Math.max(Math.abs(mechanicalNode.getTorque() / 8000d), 2);
					}
				}

				return 0;
			}

			@Override
			public boolean canConnect(ForgeDirection from, Object source)
			{
				return super.canConnect(from, source) && (from == getDirection() || from == getDirection().getOpposite());
			}

		};
	}

	@Override
	public void start()
	{
		pressureNode.reconstruct();
		super.start();
	}

	@Override
	public void invalidate()
	{
		super.invalidate();
		pressureNode.deconstruct();
	}

	@Override
	public void update()
	{
		super.update();

		if (!worldObj.isRemote && mechanicalNode.getPower() > 0)
		{
			/**
			 * Try to suck fluid in
			 */
			TileEntity tileIn = new Vector3(this).add(getDirection().getOpposite()).getTileEntity(this.worldObj);

			if (tileIn instanceof IFluidHandler)
			{
				FluidStack drain = ((IFluidHandler) tileIn).drain(getDirection(), pressureNode.maxOutput(), false);

				if (drain != null)
				{
					((IFluidHandler) tileIn).drain(getDirection(), fill(getDirection().getOpposite(), drain, true), true);
				}
			}
		}
	}

	@Override
	public int fill(ForgeDirection from, FluidStack resource, boolean doFill)
	{
		if (from == getDirection().getOpposite())
		{
			TileEntity tileOut = new Vector3(this).add(from.getOpposite()).getTileEntity(this.worldObj);

			if (tileOut instanceof IFluidHandler)
				return ((IFluidHandler) tileOut).fill(from, resource, doFill);
		}

		return 0;
	}

	@Override
	public FluidStack drain(ForgeDirection from, FluidStack resource, boolean doDrain)
	{
		return null;
	}

	@Override
	public FluidStack drain(ForgeDirection from, int maxDrain, boolean doDrain)
	{
		return null;
	}

	@Override
	public boolean canFill(ForgeDirection from, Fluid fluid)
	{
		return from == this.getDirection().getOpposite();
	}

	@Override
	public boolean canDrain(ForgeDirection from, Fluid fluid)
	{
		return from == this.getDirection();
	}

	@Override
	public FluidTankInfo[] getTankInfo(ForgeDirection from)
	{
		return null;
	}

	@Override
	public INode getNode(Class<? extends INode> nodeType, ForgeDirection from)
	{
		if (nodeType.isAssignableFrom(pressureNode.getClass()))
			return pressureNode;

		return super.getNode(nodeType, from);
	}

    @Override
    public ForgeDirection getDirection() {
        return null;
    }

    @Override
    public void setDirection(ForgeDirection direction) {

    }
}

File: src/main/scala/resonantinduction/mechanical/fluid/transport/TilePump.scala
package resonantinduction.mechanical.fluid.transport

import net.minecraft.block.material.Material
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.ResourceLocation
import net.minecraftforge.client.model.AdvancedModelLoader
import net.minecraftforge.common.util.ForgeDirection
import net.minecraftforge.fluids.{Fluid, FluidStack, FluidTankInfo, IFluidHandler}
import org.lwjgl.opengl.GL11
import resonant.api.IRotatable
import resonant.lib.render.RenderUtility
import resonant.lib.transform.vector.Vector3
import resonantinduction.core.Reference
import resonantinduction.mechanical.mech.TileMechanical

import scala.collection.mutable

object TilePump
{
  val model = AdvancedModelLoader.loadModel(new ResourceLocation(Reference.domain, Reference.modelPath + "pump.tcn"))
  val texture = new ResourceLocation(Reference.domain, Reference.modelPath + "pump.png")
}

class TilePump extends TileMechanical(Material.iron) with IRotatable with IFluidHandler
{
  val pressureNode = new PumpNode(this)

  //Constructor
  normalRender = false
  isOpaqueCube = false
  setTextureName("material_steel")

  nodes.add(pressureNode)

  override def update()
  {
    super.update()

    if (!worldObj.isRemote && mechanicalNode.getPower > 0)
    {
      //Push fluid inside this block to its front
      //TODO: Allow change of direction based on angular velocity
      val drain = pressureNode.drain(getDirection, pressureNode.getCapacity, false)

      if (drain != null)
      {
        pressureNode.drain(getDirection, fill(getDirection.getOpposite, drain, true), true)
      }

      pressureNode.maxFlowRate = Math.abs(mechanicalNode.angularVelocity * 15).toInt
    }
  }

  override def renderDynamic(pos: Vector3, frame: Float, pass: Int): Unit =
  {
    GL11.glPushMatrix()
    GL11.glTranslated(pos.x + 0.5, pos.y + 0.5, pos.z + 0.5)
    GL11.glRotatef(-90, 0, 1, 0)
    if (tile.getWorldObj != null) RenderUtility.rotateBlockBasedOnDirection(getDirection)
    RenderUtility.bind(TilePump.texture)
    val notRendered = mutable.Set.empty[String]

    GL11.glPushMatrix()
    GL11.glRotated(Math.toDegrees(mechanicalNode.angle), 0, 0, 1)

    for (i <- 1 to 12)
    {
      val fin: String = "fin" + i
      val innerFin: String = "innerFin" + i
      notRendered.add(fin)
      notRendered.add(innerFin)
      TilePump.model.renderOnly(fin, innerFin)
    }

    GL11.glPopMatrix()
    TilePump.model.renderAllExcept(notRendered.toArray[String]: _*)
    GL11.glPopMatrix()
  }

  def fill(from: ForgeDirection, resource: FluidStack, doFill: Boolean): Int =
  {
    if (from == getDirection.getOpposite)
    {
      val tileOut: TileEntity = toVector3.add(from.getOpposite).getTileEntity(this.worldObj)
      if (tileOut.isInstanceOf[IFluidHandler]) return (tileOut.asInstanceOf[IFluidHandler]).fill(from, resource, doFill)
    }
    return 0
  }

  def drain(from: ForgeDirection, resource: FluidStack, doDrain: Boolean): FluidStack = null

  def drain(from: ForgeDirection, maxDrain: Int, doDrain: Boolean): FluidStack = null

  def canFill(from: ForgeDirection, fluid: Fluid): Boolean = from == getDirection.getOpposite

  def canDrain(from: ForgeDirection, fluid: Fluid): Boolean = from == getDirection

  def getTankInfo(from: ForgeDirection): Array[FluidTankInfo] = null
}